### ISSUE TITLE
sigma vertical coords, and 3D advection

### ISSUE THREAD

**Issue by ana-machado** [OceanParcels/parcels#502]
Hello everyone,
I have some main questions about OceanParcels that were not clear for me in the tutorials.
- Is the 3D advection already working? In the tutorials I haven't seen any example reading w, only uv.
- How does parcels handle the vertical coordinate of the models? Is it possible to use outputs from a terrain-following vertical coordinate model (eg. ROMS)? If not yet, is it going to be implemented soon?
Hope anyone can clarify me :)
Best wishes,
Ana

**Comment by delandmeterp** [comment]
Hi Ana,

Yes we can do 3D advection using 3D fields, and yes we can read s-levels grids!

For the first point, 3D advection is not super visible within all the 2D examples, but it is fully validated and works well.
For the second point, it does also work and we've validated a few examples, but we don't use it so much here in the lab, so I haven't fully tested for example how it reacts to `Fieldset.from_netcdf()` function.

So please use your ROMS data and contact us in case of any problem (providing a small dataset! )

Cheers,

Philippe

**Comment by delandmeterp** [comment]
PS: Are your vertical coordinate moving in time? We can read such coordinates, but `Fieldset.from_netcdf` does not work in such case for sure. If you need this,  you can create the grid yourself following this example (line 157)
https://github.com/OceanParcels/parcels/blob/master/tests/test_grids.py

**Comment by delandmeterp** [comment]
But the sea surface level is moving, right?

**Comment by ana-machado** [comment]
Of course, yes, you are right. Sigma levels move in time with zeta. I think I am not getting what you mean with this function, line 157. Is this to read the data from roms output file?
And how do I define the depth levels of the sigmas?

**Comment by delandmeterp** [comment]
Since the depth levels are moving in time, we don't have yet a proper function to read it immediately from the files, but still, moving depths can be used with current version of Parcels.

In function starting from line 157, I show an example with sigma levels. The data are not read from the files, I do define them myself.

About how to define the depth levels of the sigmas, what do yo mean? First is to understand what is your mesh. If you have real sigma levels, it goes like this:
With n levels, the depths are
z = -h + (eta + h) * i / (n-1)
for level i (i ranging from 0 to n-1), with h the bathymetry and eta the sea surface elevation.
(or the opposite, maybe level 0 is at the surface, maybe at the bottom)
Could you check that in your data?
Then we can think how to put that into Parcels

**Comment by ana-machado** [comment]
I have the functions to define the depth levels of the sigmas, it is a bit more complex because it allows to refine the mesh near the surface and near the bottom (by changing 2 parameters, theta_s and theta_b. In the simulation I would like to use now, I have increased resolution near the surface (theta_s=6). 
z levels, where the vertical velocity is computed z_w are defined like this (matlab code):

cff1 = 1./sinh(theta_s);
cff2 = 0.5/tanh(0.5*theta_s);

sc = ((0:N) - N) / N;
N = N + 1;

Cs = (1.-theta_b) * cff1 * sinh(theta_s * sc)...
    + theta_b * (cff2 * tanh(theta_s * (sc + 0.5)) - 0.5);

hinv=1./(h+hc);
cff=hc*sc;
cff1=Cs;
for k=1:N
   z(k)=zeta+(zeta+h).*(cff(k)+cff1(k)*h).*hinv;
end 

(https://www.myroms.org/wiki/Vertical_S-coordinate)
u and v are defined at different levels (z_r) as the tracers.
 
I will really need your help to be able to do this, because now I wouldn't even know where to start.

Do you think this is going to be easy/possible?

**Comment by delandmeterp** [comment]
Yes, it is possible.

What you have to do is to create a netcdf file in which you define a depth field of dimensions
(tdim, zdim, ydim, xdim), and then you apply your formula

z[ti, zi, yi, xi]=zeta[ti, yi, xi]+(zeta[ti, yi, xi] + h[yi, xi] ).(cff[zi]+cff1[zi]*h[yi, xi]).*hinv[yi, xi];

(zi, yi, ti, xi) are the indices of the data in your arrays.

What is the dimension tdim? 
If it is big, you should maybe put one time step (ti) per netcdf file

When this is done, the Parcels part comes. You can use Fieldset.from_netcdf(...., full_load=True)
This last flag is important. Usually, we don't fully load the data since the beginning, but only the time steps which are currently necessary. All is done under the hood
But this won't work with your z moving in time.

**Comment by ana-machado** [comment]
Thank you very much for your reply. My tdim depends on the purpose, but for what I want to do first it would be around 80 days. I have outputs at each 5 days, do you think it is enough to have a credible results?

I will try to do the first part, and I will get back to you after that.

**Comment by delandmeterp** [comment]
With tdim around 20, you could potentially use this full_load=True. (it depends also on the other dimensions, and the RAM of your computer, but you can definitely try).
This means that it will be easier to work once we have the zdepth in a netcdf file, good news!

**Comment by ana-machado** [comment]
Sorry, I meant, 80 time steps, not 80 days (one year). So probably it is too much, then. I should make a netcdf file for each step, in this case? And shall I keep the "time" dimension, anyway?

**Comment by delandmeterp** [comment]
80 time steps will depend on the xdim, ydim and zdim. We'll try and see.
It doesn't really matter if you put it in one single file or in multiple files.
It is good to have a time dimension and time value attached to your different depths (both if you use a single file or multiple files)

**Comment by ana-machado** [comment]
ok. thank you :) I will try and write you back when I have made some progress.

**Comment by ana-machado** [comment]
Hello, Happy New Year!
I have already prepared the roms files adding them 2 new variables:
	float SigmaDepths(s_rho, eta_rho, xi_rho) ;
	float SigmaDepthsVar(time, s_rho, eta_rho, xi_rho) ;
They are the average depth of the sigmas (SigmaDepths) and the depth of the sigmas varying in time (SigmaDepthsVar).
What do you suggest me to do now? I thought I could start to try to run parcels considering the depths averaged in time. The results shouldn't be much different and it would probably simplify the process a lot, no?
In any case, I am not sure what should I do next. Do you have any example you can send me?
Thank you very much
Ana

My netcdf structure:
netcdf roms_avg_Y2000_SigmasTime.nc {
dimensions:
	eta_rho = 482 ;
	xi_rho = 368 ;
	time = UNLIMITED ; // (73 currently)
	s_rho = 40 ;
	xi_u = 367 ;
	eta_v = 481 ;
variables:
	float h(eta_rho, xi_rho) ;
		h:long_name = "bathymetry at RHO-points" ;
		h:units = "meter" ;
		h:field = "bath, scalar" ;
	float lat_rho(eta_rho, xi_rho) ;
		lat_rho:long_name = "latitude of RHO-points" ;
		lat_rho:units = "degree_north" ;
		lat_rho:field = "lat_rho, scalar" ;
	float lon_rho(eta_rho, xi_rho) ;
		lon_rho:long_name = "longitude of RHO-points" ;
		lon_rho:units = "degree_east" ;
		lon_rho:field = "lon_rho, scalar" ;
	float mask_rho(eta_rho, xi_rho) ;
		mask_rho:long_name = "mask on RHO-points" ;
		mask_rho:option_0 = "land" ;
		mask_rho:option_1 = "water" ;
	float pm(eta_rho, xi_rho) ;
		pm:long_name = "curvilinear coordinate metric in XI" ;
		pm:units = "meter-1" ;
		pm:field = "pm, scalar" ;
	float pn(eta_rho, xi_rho) ;
		pn:long_name = "curvilinear coordinate metric in ETA" ;
		pn:units = "meter-1" ;
		pn:field = "pn, scalar" ;
	float salt(time, s_rho, eta_rho, xi_rho) ;
		salt:long_name = "averaged salinity" ;
		salt:units = "PSU" ;
		salt:field = "salinity, scalar, series" ;
	float scrum_time(time) ;
		scrum_time:long_name = "averaged time since initialization" ;
		scrum_time:units = "second" ;
		scrum_time:field = "time, scalar, series" ;
	float sustr(time, eta_rho, xi_u) ;
		sustr:long_name = "averaged Kinematic u wind stress component" ;
		sustr:units = "N/m2" ;
	float svstr(time, eta_v, xi_rho) ;
		svstr:long_name = "averaged Kinematic v wind stress component" ;
		svstr:units = "N/m2" ;
	float temp(time, s_rho, eta_rho, xi_rho) ;
		temp:long_name = "averaged potential temperature" ;
		temp:units = "Celsius" ;
		temp:field = "temperature, scalar, series" ;
	float u(time, s_rho, eta_rho, xi_u) ;
		u:long_name = "averaged u-momentum component" ;
		u:units = "meter second-1" ;
		u:field = "u-velocity, scalar, series" ;
	float v(time, s_rho, eta_v, xi_rho) ;
		v:long_name = "averaged v-momentum component" ;
		v:units = "meter second-1" ;
		v:field = "v-velocity, scalar, series" ;
	float w(time, s_rho, eta_rho, xi_rho) ;
		w:long_name = "averaged vertical momentum component" ;
		w:units = "meter second-1" ;
		w:field = "w-velocity, scalar, series" ;
	float zeta(time, eta_rho, xi_rho) ;
		zeta:long_name = "averaged free-surface" ;
		zeta:units = "meter" ;
		zeta:field = "free-surface, scalar, series" ;
	float SigmaDepths(s_rho, eta_rho, xi_rho) ;
	float SigmaDepthsVar(time, s_rho, eta_rho, xi_rho) ;

**Comment by delandmeterp** [comment]
Hi Ana,

I'm currently working at sea with very poor access to internet. I'll be back in February when it will be easier to help you.

If your mesh doesn't move too much vertically, a first approximation with the mean position should be a good start. To do so, you can use `Fieldset.from_netcdf()` with in the arguments the dictionaries giving the variables names and file names for each dimension (depth, lon, lat and the data).

I don't have an exact example, but in this repo, you can find a similar use with `Fieldset.from_nemo` but you can use `from_netcdf` if your grid is a A-grid. Don't forget to add the depth dimension (not in the example)
https://github.com/OceanParcels/Parcelsv2.0PaperNorthSeaScripts

**Comment by ana-machado** [comment]
Hi Philippe,
Thank you very much for your reply. I will try to do a simulation with the mean sigmas and contact you back in February.
Cheers,
Ana

**Comment by ana-machado** [comment]
Hi Philippe, or someone who could help,
I am trying to prepare a configuration to use with roms model, that has sigma vertical coordinates. I am considering they don't change in time, in a first approach. I am having this error:

 --> 174                     raise NotImplementedError('Vertically adaptive meshes not implemented for from_netcdf()')

Is there something I am missing? My code is below.

Ana


    filenames = {'U': {'depth': files, 'lon': grdfile, 'lat': grdfile, 'data': files},
                 'V': {'depth': files, 'lon': grdfile, 'lat': grdfile, 'data': files}}

    variables = {'U': 'u',
                 'V': 'v'}
    dimensions = {'U': {'time': 'scrum_time', 'depth': 'SigmaDepths', 'lat': 'lat_u', 'lon': 'lon_u'},
                  'V': {'time': 'scrum_time', 'depth': 'SigmaDepths', 'lat': 'lat_u', 'lon': 'lon_u'}}

    fieldset = FieldSet.from_nemo(filenames, variables, dimensions)

**Comment by ana-machado** [comment]
Hi Philippe,
Not sure you're already back, if you can help me with this it would be great! Right now I have no idea what to do next.
Cheers,
Ana

**Comment by delandmeterp** [comment]
Hi Ana,
I guess the codes thinks you want to give him a vertically adaptive mesh, since you give him multiple files for the depth value. Since the depth is constant in time, you should give only one single file.
Does that fix your problem?

Cheers,
Philippe

**Comment by ana-machado** [comment]
Hi Philippe,
Thank you for your reply. I am not sure I am doing your suggestion, I changed for this:
    filenames = {'U': {'depth': file1, 'lon': grdfile, 'lat': grdfile, 'data': files},
                 'V': {'depth': file1, 'lon': grdfile, 'lat': grdfile, 'data': files}}

    variables = {'U': 'u',
                 'V': 'v'}
    dimensions = {'U': {'time': 'scrum_time', 'depth': 'SigmaDepths', 'lat': 'lat_u', 'lon': 'lon_u'},
                  'V': {'time': 'scrum_time', 'depth': 'SigmaDepths', 'lat': 'lat_u', 'lon': 'lon_u'}}


filenames output is:
{'U': {'data': ['./ncfiles/roms_avg_Y2000_interm.nc.1'],
  'depth': './ncfiles/roms_avg_Y2000_interm.nc.1',
  'lat': './ncfiles/roms_grd.nc.1',
  'lon': './ncfiles/roms_grd.nc.1'},
 'V': {'data': ['./ncfiles/roms_avg_Y2000_interm.nc.1'],
  'depth': './ncfiles/roms_avg_Y2000_interm.nc.1',
  'lat': './ncfiles/roms_grd.nc.1',
  'lon': './ncfiles/roms_grd.nc.1'}}


it looks ok, no?

but I'm still getting the error:

In [16]:     fieldset = FieldSet.from_nemo(filenames, variables, dimensions)
    ...: 
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-16-164addbaf797> in <module>()
----> 1 fieldset = FieldSet.from_nemo(filenames, variables, dimensions)

/home/ana/anaconda2/envs/py2_parcels/lib/python2.7/site-packages/parcels/fieldset.pyc in from_nemo(cls, filenames, variables, dimensions, indices, mesh, allow_time_extrapolation, time_periodic, tracer_interp_method, **kwargs)
    292 
    293         return cls.from_netcdf(filenames, variables, dimensions, mesh=mesh, indices=indices, time_periodic=time_periodic,
--> 294                                allow_time_extrapolation=allow_time_extrapolation, interp_method=interp_method, **kwargs)
    295 
    296     @classmethod

/home/ana/anaconda2/envs/py2_parcels/lib/python2.7/site-packages/parcels/fieldset.pyc in from_netcdf(cls, filenames, variables, dimensions, indices, mesh, allow_time_extrapolation, time_periodic, full_load, **kwargs)
    229             fields[var] = Field.from_netcdf(paths, var, dims, inds, grid=grid, mesh=mesh,
    230                                             allow_time_extrapolation=allow_time_extrapolation,
--> 231                                             time_periodic=time_periodic, full_load=full_load, **kwargs)
    232         u = fields.pop('U', None)
    233         v = fields.pop('V', None)

/home/ana/anaconda2/envs/py2_parcels/lib/python2.7/site-packages/parcels/field.pyc in from_netcdf(cls, filenames, variable, dimensions, indices, grid, mesh, allow_time_extrapolation, time_periodic, full_load, **kwargs)
    172                 depth_filename = filenames['depth'][0]
    173                 if len(depth_filename) != 1:
--> 174                     raise NotImplementedError('Vertically adaptive meshes not implemented for from_netcdf()')
    175         else:
    176             lonlat_filename = filenames[0]

NotImplementedError: Vertically adaptive meshes not implemented for from_netcdf()

In [17]:

**Comment by ana-machado** [comment]
I am also getting an error message when I try to import NestedField:
In [17]: from parcels import  NestedField
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-17-4a575e4899b1> in <module>()
----> 1 from parcels import  NestedField

ImportError: cannot import name NestedField

**Comment by delandmeterp** [comment]
could you send me a few (small) datafiles such that I can give it a try?

The second problem (with the NestedField import) must be something with your Parcels version. How did you update Parcels? Maybe part of the code already uses NestedField, but not another?

**Comment by ana-machado** [comment]
Regarding the second question, I didn't update. I am using the version I downloaded and this function you sent me:
https://github.com/OceanParcels/Parcelsv2.0PaperNorthSeaScripts
the code was probably updated in between. Is there any way to update, or shall I download the new code and follow the installation instructions again?
I uploaded the files I am trying to use to this directory:
https://meocloud.pt/link/eef0a305-0600-474f-8a7c-837017349702/var/
they are not much small, if they are too big for you let me know, and I'll try to create another with fewer time steps
thank you very much for your help
Ana

**Comment by ana-machado** [comment]
Hi Philippe,
Did you manage to try anything with the files I sent you?
Best,
Ana

**Comment by delandmeterp** [comment]
Hi Ana,

The problem with the NestedField is that you are using my code which import NestedField, but your parcels version does not have them. Just remove the `import NestedField` line

I've tried to download the data now, but the downloading will last 5hours according to my internet browser. Could you create a netcdf file with a few 10Mb (max 500Mb) ? Thanks!

**Comment by ana-machado** [comment]
Hi Philippe,
I kept only 4 time steps and the velocity and grid variables, the file is out.nc in the same directory:
https://meocloud.pt/link/eef0a305-0600-474f-8a7c-837017349702/var/
Thank you very much
Ana

**Comment by delandmeterp** [comment]
Hi Ana,
I wanted to use the nc files to load them into a Parcels fieldset, but variables are missing.
On your post of Jan 15, you were trying to import variable `SigmaDepths`, but I don't see it in the files.

I can help you to do the first step of loading the nc files into a Parcels fieldset, but I need for that all the information in the ncfiles. I won't have time to create the depth layers as you were explaining on Dec 11th. Could you please provide me a comprehensive nc files containing the field dimensions?

Thank you!

**Comment by ana-machado** [comment]
Hi Phillipe,
You are totally right, I am sorry about the mess. In the process of reducing the size, I forgot to add the variables with the depth of the sigmas. I was about to send you the correct files yesterday, but checking if everything was fine this time, I noticed the files were corrupted and had some weird velocity values. I decided to re-do all of the process and its finally done. I just uploaded a new out.nc 
file with (I hope) everything necessary (https://meocloud.pt/link/eef0a305-0600-474f-8a7c-837017349702/var/. 
eta_tho is y dimension
xi_rho is x dimension
eta_v is y dimension of meridional velocity
xi_u is u dimension of zonal velocity
s_rho is the number of sigma vertical levels

The variables: - SigmaDepths is the average sigma depths (computed with average zeta (free surface))
                         - SigmaDepthsVar is sigma depths varying in time (computed with zeta(t))

Thank you very much for your availability. 
Hope it works! :)
Ana

**Comment by delandmeterp** [comment]
Dear Ana,

I have finally found the time to download your data and process them. I'm sorry for the delay. As you've seen, we are very busy with the general development of the code and our own ocean applications and can't afford to spend time on all users applications.

As you will see in the example attached, Parcels works smoothly with the vertical coordinates you provided. Also, in your mesh, u, v, w and tracer fields are not provided on the same grid. It is an example of a C-grid.
Such grid is also processed by Parcels, but this requires some assumption in the indexing of the gridded data. Fortunately it is the case in the data you provided, except that your u velocity has one row too much, v velocity has one column too much, and the tracers and w velocity have one row and one column too much. But the unnecessary data are ignored by the code.
To read more about those issues, please read the following submitted paper (https://www.geosci-model-dev-discuss.net/gmd-2018-339/) 
You can see in cell number 5 of the attached notebook that your indexing is the same as in the paper (fig. 2)

Also, I haven't checked in your data what w velocity means.
If increasing w means increasing z, everything is ok. Otherwise (like e.g. in nemo, where w is upward, but when going upward, z decreases), you should use `fieldset.W.set_scaling_factor(-1)`.

So basically, you will need to use only cell 13 of the notebook, but please read all the cells, to understand the structure of the data you are handling.

Final comment, you see in that cell this line `fieldset.U.grid.time_full = fieldset.U.grid.time_full.astype(np.float64)`. This is due to a bug that I've just found and will fix asap...


[Test_ana_dataset.pdf](https://github.com/OceanParcels/parcels/files/2953650/Test_ana_dataset.pdf)
Cheers,
Philippe
